cmd/server/optimus.go

```diff
@@ -0,0 +1,347 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	"github.com/odpf/salt/log"
+	"github.com/prometheus/client_golang/prometheus"
+	slackapi "github.com/slack-go/slack"
+	"google.golang.org/grpc"
+	"gorm.io/gorm"
+
+	v1handler "github.com/odpf/optimus/api/handler/v1beta1"
+	pb "github.com/odpf/optimus/api/proto/odpf/optimus/core/v1beta1"
+	"github.com/odpf/optimus/config"
+	"github.com/odpf/optimus/datastore"
+	"github.com/odpf/optimus/ext/notify/slack"
+	"github.com/odpf/optimus/job"
+	"github.com/odpf/optimus/models"
+	"github.com/odpf/optimus/run"
+	"github.com/odpf/optimus/service"
+	"github.com/odpf/optimus/store/postgres"
+	"github.com/odpf/optimus/utils"
+)
+
+type setupFn func() error
+
+type OptimusServer struct {
+	conf   config.Optimus
+	logger log.Logger
+
+	appKey models.ApplicationKey
+	dbConn *gorm.DB
+
+	serverAddr string
+	grpcServer *grpc.Server
+	httpServer *http.Server
+
+	cleanupFn []func()
+}
+
+func New(l log.Logger, conf config.Optimus) (*OptimusServer, error) {
+	addr := fmt.Sprintf("%s:%d", conf.Server.Host, conf.Server.Port)
+	server := &OptimusServer{
+		conf:       conf,
+		logger:     l,
+		serverAddr: addr,
+	}
+
+	if err := checkRequiredConfigs(conf.Server); err != nil {
+		return server, err
+	}
+
+	fns := []setupFn{
```

Comment:
rename the fns as setupFNs and move startListening outside and call it explicitly